### PR TITLE
CB-8641 Some file-transfer plugin tests occasionally fail in mobilesp…

### PR DESCRIPTION
**Pull request by daserge**
…ec on Windows Phone 8.1

Increases timeouts for Windows platform in abort tests
Adds implementation for upload operation on Windows to abort right away rather than actually performing the upload

[Jira issue](https://issues.apache.org/jira/browse/CB-8641)

### CODE REVIEW

File: src/windows/FileTransferProxy.js [hunk 1]
```diff
@@ -116,9 +118,13 @@ exec(win, fail, 'FileTransfer', 'upload',
 
             // check if download isn't already cancelled
             var uploadOp = fileTransferOps[uploadId];
-            if (uploadOp && uploadOp.state === FileTransferOperation.CANCELLED) {
+            if (uploadOp && uploadOp.state === FileTransferOperation.CANCELLED || abortedOpIds.indexOf(uploadId) !== -1) {
```
Comment (dblotsky) [inline src/windows/FileTransferProxy.js]: This condition is hard to read. Please encapsulate this logic in a descriptive function.

File: src/windows/FileTransferProxy.js [hunk 2]
```diff
@@ -152,6 +158,18 @@ exec(win, fail, 'FileTransfer', 'upload',
             try {
                 uploader.createUploadAsync(uri, transferParts).then(
                     function (upload) {
+                        // check if download isn't already cancelled
+                        var uploadOp = fileTransferOps[uploadId];
+                        if (uploadOp && uploadOp.state === FileTransferOperation.CANCELLED || abortedOpIds.indexOf(uploadId) !== -1) {
```
Comment (dblotsky) [inline src/windows/FileTransferProxy.js]: Is it possible to factor out and reuse the code from above here?

File: tests/tests.js [hunk 3]
```diff
@@ -468,6 +468,8 @@ exports.defineAutoTests = function () {
                 it("filetransfer.spec.10 should be stopped by abort() right away", function (done) {
 
                     var fileURL = 'http://cordova.apache.org/downloads/BlueZedEx.mp3';
+                    // Some Windows Phone 8.1 devices need ~1-2 sec to create an upload operation.
+                    var graceTimeout = isWindows ? GRACE_TIME_DELTA * 6 : GRACE_TIME_DELTA;
```
Comment (dblotsky) [inline tests/tests.js]: This might be best set at the top of the file, along with the other constants.

File: tests/tests.js [hunk 4]
```diff
@@ -450,7 +450,7 @@ exports.defineAutoTests = function () {
                         expect(transfer.onprogress).toHaveBeenCalled();
 
                         // check that there is no file
-                        root.getFile(localFilePath, null, unexpectedCallbacks.fileSystemWin, done);
+                        root.getFile(fileName, null, unexpectedCallbacks.fileSystemWin, done);
```
Comment (dblotsky) [inline tests/tests.js]: Why was this changed?

**Comment by daserge** [comment]
@dblotsky, could you please review once more?

File: README.md [hunk 5]
```diff
@@ -230,6 +230,10 @@ __Parameters__:
 
 - Download requests is being cached by native implementation. To avoid caching, pass `if-Modified-Since` header to download method.
 
+### Windows Quirks
+
+- Usage of [BackgroundDownloader](https://msdn.microsoft.com/en-us/library/windows/apps/windows.networking.backgroundtransfer.backgrounddownloader.aspx) entails the latency issues on Windows Phone. You can use XHR or [HttpClient](https://msdn.microsoft.com/en-us/library/windows/apps/windows.web.http.httpclient.aspx) as a quicker alternative for small downloads.
```
Comment (jasongin) [inline README.md]: Just "Windows", not "Windows Phone". I'm pretty sure this is not a phone-only issue since the APIs are the same on Windows desktop.

Also, is the reader going to understand what the "latency issues" are, or does it need more explanation here?

File: tests/tests.js [hunk 6]
```diff
@@ -29,6 +29,7 @@ exports.defineAutoTests = function () {
 
     // constants
     var GRACE_TIME_DELTA = 600; // in milliseconds
+    var WINDOWS_GRACE_TIME_DELTA = 3500; // in milliseconds; NOTE: Some Windows Phone 8.1 devices need a few seconds to create an upload operation.
```
Comment (jasongin) [inline tests/tests.js]: Are you sure it is only WP 8.1? What about Windows 10 mobile or desktop?

File: src/windows/FileTransferProxy.js [hunk 7]
```diff
@@ -49,6 +49,17 @@ function nativePathToCordova(path) {
     return String(path).replace(/\\/g, '/');
 }
 
+function alreadyCancelled(opId, errorCb, filePath, server) {
+    var op = fileTransferOps[opId];
+    if (op && op.state === FileTransferOperation.CANCELLED) {
+        // Here we should call errorCB with ABORT_ERR error
+        errorCb(new FTErr(FTErr.ABORT_ERR, nativePathToCordova(filePath), server));
```
Comment (dblotsky) [inline src/windows/FileTransferProxy.js]: This side effect should not be part of the function. Please place it outside, into the calling code. It will be duplicated, but that is still somewhat OK; it's definitely better than someone calling if `alreadyCancelled()` twice in a row and getting their callback called twice.

Comment (daserge) [inline src/windows/FileTransferProxy.js]: Addressed, thanks!

File: README.md (hunk 5 above)
Comment (daserge) [inline README.md]: Moved quirk out from `download` method section as it affects uploads as well; added some details on the latency issue.

**Comment by daserge** [comment]
@dblotsky, @jasongin, I've addressed the CR notes and also added a few fixes for the last commit changed the timeouts - please take a look.

**Comment by omefire** [comment]
The mention of BackgroundUploader/Downloaded in the Windows quirks section confused me a bit.
At first, I thought they were some options we could use to configure the upload/download operations.
Could we maybe mention that it is an implementation detail in Windows land ? And make it a bit clearer what the alternatives are ?

**Comment by jasongin** [comment]
LGTM

File: tests/tests.js [hunk 8]
```diff
@@ -31,9 +31,11 @@ exports.defineAutoTests = function () {
     var ONE_SECOND = 1000; // in milliseconds
     var GRACE_TIME_DELTA = 600; // in milliseconds
     var DEFAULT_FILESYSTEM_SIZE = 1024 * 50; // filesystem size in bytes
+    var WINDOWS_GRACE_TIME_DELTA = 5 * ONE_SECOND; // Some Windows devices need a few seconds to create an upload/download operation.
     var UNKNOWN_HOST = "http://foobar.apache.org";
     var HEADERS_ECHO = "http://whatheaders.com"; // NOTE: this site is very useful!
     var DOWNLOAD_TIMEOUT = 7 * ONE_SECOND;
+    var WINDOWS_UNKNOWN_HOST_TIMEOUT = 30 * ONE_SECOND;
```
Comment (dblotsky) [inline tests/tests.js]: What's this for?

Comment (daserge) [inline tests/tests.js]: The filetransfer.spec.15 takes 20-30 seconds to finish so 7 seconds is just not enough.
